lib/rust/flexer/src/group.rs

```diff
@@ -10,6 +10,150 @@ pub mod rule;
+        let err = format!("The provided group_id {:?} is invalid.",group);
+        let group = self.group_from_id_mut(group).expect(&err);
+        group.add_rule(rule);
+    }
+
+    /// Collates the entire set of rules that are matchable when the lexer has the group identified
+    /// by `group_id` as active.
+    ///
+    /// This set of rules includes the rules inherited from any parent groups.
+    pub fn rules_for(&self, group:Identifier) -> Option<Vec<&Rule>> {
+        self.group_from_id(group).map(|group| {
+            let mut parent = group.parent_index.and_then(|ix|self.group_from_id(ix.into()));
+            let mut rules = (&group.rules).iter().collect_vec();
+            while let Some(parent_group) = parent {
+                if parent_group.id == group.id {
+                    panic!("There should not be cycles in parent links for lexer groups.")
+                }
+                rules.extend((&parent_group.rules).iter());
+                parent = parent_group.parent_index.and_then(|ix|self.group_from_id(ix.into()));
+            }
+
+            rules
+        })
+    }
+
+    /// Obtains a reference to the group for the given `group_id`.
+    pub fn group_from_id(&self, group:Identifier) -> Option<&Group> {
+        self.groups.get(group.val)
+    }
+
+    /// Obtains a mutable reference to the group for the given `group_id`.
+    pub fn group_from_id_mut(&mut self, group:Identifier) -> Option<&mut Group> {
+        self.groups.get_mut(group.val)
+    }
+
+    /// Converts the group identified by `group_id` into an NFA.
+    ///
+    /// Returns `None` if the group does not exist, or if the conversion fails.
+    pub fn to_nfa_from(&self, group:Identifier) -> Option<NFA> {
+        let group = self.group_from_id(group);
+        group.map(|group| {
+            let mut nfa = NFA::default();
+            let start   = nfa.new_state();
+            let build   = |rule:&Rule| nfa.new_pattern(start,&rule.pattern);
+            let rules   = self.rules_for(group.id).expect("Group exists.");
+            let states  = rules.into_iter().map(build).collect_vec();
+            let end     = nfa.new_state();
+            for (ix,state) in states.into_iter().enumerate() {
+                nfa.states[state.id].name = Some(group.callback_name(ix));
+                nfa.connect(state,end);
+            }
+            nfa
+        })
+    }
+
+    /// Generates the next group identifier for this registry.
+    fn next_id(&self) -> usize {
+        self.groups.len()
+    }
+}
+
+
+
+// ==================
+// === Identifier ===
+// ==================
+
+/// An identifier for a group.
+#[allow(missing_docs)]
+#[derive(Copy,Clone,Debug,Default,Eq,PartialEq)]
+pub struct Identifier {
+    val:usize
+}
+
+impl Identifier {
+    /// Creates a new identifier.
+    pub fn new(val:usize) -> Identifier {
+        Identifier{val}
+    }
+}
+
+
+// === Trait Impls ===
+
+impl From<usize> for Identifier {
```

Comment:
It's always worth adding `From<&usize>` as well. It is often use for iterators etc and justshouldb e defined.